File: Leetcode/Javascript/48. Rotate Image.js
var rotate = function(matrix) {
    //-----diagonal-----//
    var row = matrix.length;
    if(!row) return;
    //-----diagonal-----//
    for(var i = 0; i < row-1; ++i) {
        for(var j = 0; j < row-1-i; ++j) {
            var tmp = matrix[i][j];
            matrix[i][j] = matrix[row-1-j][row-1-i];
            matrix[row-1-j][row-1-i] = tmp;
        }
    }
    //------------------//
    for(i = 0; i < (row >> 1);++i) {
        for(j = 0; j < row; ++j) {
            tmp = matrix[i][j];
            matrix[i][j] = matrix[row-1-i][j];
            matrix[row-1-i][j] = tmp;
        }
    }
};

File: Leetcode/One/48. Rotate Image.java
class Solution {
    public void rotate(int[][] matrix) {
        //up down
        int height = matrix.length,
            width = matrix[0].length;
        int halfHeight = height >> 1;
        for(int i = 0; i < halfHeight; ++i) {
            for(int j = 0; j < width; ++j) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[height-1-i][j];
                matrix[height-1-i][j] = tmp;
            }
        }
        //left diag
        for(int i = 0; i < height; ++i) {
            for(int j = i; j < width; ++j) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }
    }
}
